### PR TITLE
K8s native binding design proposal

### PR DESCRIPTION
**Pull request by RedbackThomson**
Issue #, if available: #740

Description of changes:
Adds a proposal for ACK-K8s native binding, allowing for K8s applications to bind to ACK resource fields.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/design/proposals/native-binding/native-binding.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+### Redhat Service Binding Operator
+
+>“[Redhat Service Binding Operator] enables developers to connect their application to backing services with a consistent and predictable experience“ - https://github.com/redhat-developer/service-binding-operator
+
+
+The [Redhat Service Binding Operator](https://github.com/redhat-developer/service-binding-operator) (herein called the SBO) uses annotations present on CRs and resources to import field values from other resources. The main purpose of SBO is to standardise the system for injecting secret information from database services into application deployment environment variables. The SBO can automatically detect and bind to services only from a limited number of [supported operators](https://github.com/redhat-developer/service-binding-operator#known-bindable-operators). 
+
+SBO is not suitable for this design as it can only support a limited number of fields based on pre-configured database bindings. That is, it can handle the case where an application developer wants to inject a PostgreSQL database connection string into a deployment, but little more than that. SBO does not support exporting to configmap or secrets directly.
```
[the remark this answers is not in the document]
Comment (RedbackThomson) [reply docs/design/proposals/native-binding/native-binding.md:30]: > If this is not supported generally, perhaps this can be supported from the RDS (and possibly other DB service) controller?

It absolutely could be, but not included as part of the implementation for this more general-purpose issue. I have not looked into what it would take to add our CRDs to RSBO.

File: docs/design/proposals/native-binding/native-binding.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+This solution would introduce a new CRD solely for the purpose of declaring field exports. The CRD spec would contain an array of structs, each containing:
+
+* The source path, containing:
+    * Resource API group, kind, namespace, name
+* Spec/Status field path
+* An output paths, containing:
+    * Type: ConfigMap or Secret
+    * Namespace and name
+    * (Optionally) Output structure
```
[the remark this answers is not in the document]
Comment (RedbackThomson) [reply docs/design/proposals/native-binding/native-binding.md:46]: > ready made credential available without additional configuration 

How would you envision this working? AFAIK credentials are typically ingested through environment variables or volumes mounted directly to the pod. This solution aimed to make it easy to mount them as environment variables, by referencing a generated configmap/secret. 

Are you suggesting that we provide preformatted connection URLs as the output?

[the remark this answers is not in the document]
Comment (jkatz) [reply docs/design/proposals/native-binding/native-binding.md:46]: @RedbackThomson 

> How would you envision this working? AFAIK credentials are typically ingested through environment variables or volumes mounted directly to the pod. This solution aimed to make it easy to mount them as environment variables, by referencing a generated configmap/secret.

We're on the same page here. My thought is having the Secret/ConfigMap automatically generated by a specific controller when we already know the format of the credentials.

For example, let's use ACK for RDS. We know the [username / password](https://aws-controllers-k8s.github.io/community/docs/tutorials/rds-example/#postgresql) because that is part of custom resource creation. We will get the [host/port](https://aws-controllers-k8s.github.io/community/docs/tutorials/rds-example/#connect-to-database-instances) during the RDS provisioning process.

From a RDS perspective, these are "well-known" attributes, and it would simplify things for users if we are able to create the initial Secret with the credentials for them.

Perhaps in this case, the controller could generate an ACK Binding CRD?

File: docs/design/proposals/native-binding/native-binding.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+This design aims to address two separate use cases:
+
+1. An application team manages their own ACK resources and K8s deployments. They wish to dynamically link the status of an ACK resource to the environment variables of their deployment.
+2. An infrastructure team manages the ACK resources on behalf of application teams. The application teams do not have read/write access to the ACK resource directly, but require access to the fields from them in order to deploy their application.
```
Comment (jaypipes) [inline docs/design/proposals/native-binding/native-binding.md:12]: Ah, yes, what I said above :)

File: docs/design/proposals/native-binding/native-binding.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+## Requirements
+
+1. Export any number of Spec and/or Status values to application consumable formats (ConfigMap and Secrets)
```
Comment (jaypipes) [inline docs/design/proposals/native-binding/native-binding.md:16]: I'd reword this to:

> Allow the cluster operator to configure an ACK controller to write the value of a custom resource Spec or Status field to either a ConfigMap or a Secret

File: docs/design/proposals/native-binding/native-binding.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    * Resource API group, kind, namespace, name
+* Spec/Status field path
+* An output paths, containing:
+    * Type: ConfigMap or Secret
+    * Namespace and name
+    * (Optionally) Output structure
+
+The common field exporter reconciler running within the controller would have a shared informer for all custom resources of this type. The reconciler would read the spec for each binding resource, filter for any resources where the source path matches its respective resource API group, version and kind and then read and export the corresponding fields. 
+
+Having this configuration as a separate CRD allows application developers to specify their own configuration independent of the ops teams that manage the infrastructure. Cluster operators can lock down application teams’ RBAC permissions to only have access to read ACK resources, with an exception only to write this new binding CRD.
```
[the remark this answers is not in the document]
Comment (RedbackThomson) [reply docs/design/proposals/native-binding/native-binding.md:50]: > Couldn't the application developer circumvent the lack of `read` permissions on the CR by simply creating a bunch of `FieldExport` CRs that essentially exported all fields in the CR? :(

Yeah ... this did come up in the design review. We really didn't want to go down the route of reinventing an allow-/block-listing configuration because that would basically be reimplementing RBAC. Instead, we decided to limit this design so that you can only read fields from resources in the same namespace. We assume that if the developer has been given access to a namespace that they could be allowed to read anything within it. 

> I actually don't think it's going to be common for cluster operators to give write permission on the FieldExport CR when they don't give read permissions to the application developer role to begin with

I agree. My use case above addresses the case where the cluster ops team doesn't permit *write* access. In which case the application developer couldn't patch an annotation onto the resource to configure the exporting - hence requiring a separate CR. I am assuming for most of this that the app dev has read access.

Something that I haven't put in this doc that I just thought about is how to handle exporting fields that have secrets attached. In my mind we would not allow anyone to export those fields. Partly because it would require resolving the field and doing a lookup, which requires more RBAC permissions. Mostly though because if you need that secret in your application, you should already be able to just mount the secret as an environment variable directly, rather than exporting it and then mounting it.

File: docs/design/proposals/native-binding/native-binding.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+As an example:
+
+```yaml
+apiVersion: services.k8s.aws/v1alpha1
+kind: FieldExport
+metadata:
+  name: export-vpc-id
+spec:
+  to:
+    type: ConfigMap
```
[the remark this answers is not in the document]
Comment (RedbackThomson) [reply docs/design/proposals/native-binding/native-binding.md:61]: Sure, yeah. `type` is a reserved keyword in Go anyway, so this might conflict anyway. Probably better to keep it separate.

File: docs/design/proposals/native-binding/native-binding.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (RedbackThomson) [reply docs/design/proposals/native-binding/native-binding.md:46]: > My thought is having the Secret/ConfigMap automatically generated by a specific controller when we already know the format of the credentials

This scares me. I wouldn't want to dump database credentials into K8s without someone explicitly allowing it. I'm thinking of a case where someone has permissions to decrypt secrets within a namespace and can now suddenly see the database password of every RDS database in that namespace. 

I would like to eventually publish some templates for creating an RDS database, including all VPC resources, as a Helm chart. I think it would be great to include this exported field as a `exportCredentials: true/false` value in that chart, though.

File: docs/design/proposals/native-binding/native-binding.md [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+secret information from database services into application deployment
+environment variables. The SBO can automatically detect and bind to services
+only from a limited number of [supported
+operators](https://github.com/redhat-developer/service-binding-operator#known-bindable-operators). 
+
+SBO is not suitable for this design as it can only support a limited number of
+fields based on pre-configured database bindings. That is, it can handle the
+case where an application developer wants to inject a PostgreSQL database
+connection string into a deployment, but little more than that. SBO does not
+support exporting to configmap or secrets directly.
```
Comment (acornett21) [inline docs/design/proposals/native-binding/native-binding.md:60]: > SBO does not support exporting to configmap or secrets directly.

I might be reading their documentation backwards or not understanding it fully, but I feel it supports configmaps and secrets. Am I reading the [documentation](https://redhat-developer.github.io/service-binding-operator/userguide/exposing-binding-data/olm-descriptors.html) incorrectly?

**Comment by RedbackThomson** [comment]
Implemented with https://github.com/aws-controllers-k8s/runtime/pull/75 and https://github.com/aws-controllers-k8s/code-generator/pull/307
